modules/ballerina-core/src/main/java/org/ballerinalang/util/semantics/SemanticAnalyzer.java

```diff
@@ -1173,18 +1193,46 @@ public void visit(VariableDefStmt varDefStmt) {
     @Override
     public void visit(AssignStmt assignStmt) {
         Expression[] lExprs = assignStmt.getLExprs();
+        if (assignStmt.isVarDeclaration()) {
+            for (Expression expr : lExprs) {
+                if (!(expr instanceof VariableRefExpr)) {
+                    BLangExceptionHelper.throwSemanticError(assignStmt, SemanticErrors.INVALID_VAR_ASSIGNMENT);
+                }
+                VariableRefExpr refExpr = (VariableRefExpr) expr;
+                Identifier identifier = new Identifier(refExpr.getVarName());
+                SymbolName symbolName = new SymbolName(identifier.getName());
+                VariableDef variableDef = new VariableDef(refExpr.getNodeLocation(),
+                        refExpr.getWhiteSpaceDescriptor(), identifier,
+                        new SimpleTypeName("var"), symbolName, currentScope);
+                variableDef.accept(this);
+            }
+        }
         visitLExprsOfAssignment(assignStmt, lExprs);
 
         Expression rExpr = assignStmt.getRExpr();
         if (rExpr instanceof FunctionInvocationExpr || rExpr instanceof ActionInvocationExpr) {
             rExpr.accept(this);
+            if (assignStmt.isVarDeclaration()) {
+                if (rExpr instanceof FunctionInvocationExpr) {
+                    assignVariableRefTypes(lExprs, ((FunctionInvocationExpr) rExpr).getTypes());
+                } else if (rExpr instanceof ActionInvocationExpr) {
+                    assignVariableRefTypes(lExprs, ((ActionInvocationExpr) rExpr).getTypes());
+                }
+            }
             checkForMultiAssignmentErrors(assignStmt, lExprs, (CallableUnitInvocationExpr) rExpr);
             return;
         }
 
         if (lExprs.length > 1 && (rExpr instanceof TypeCastExpression || rExpr instanceof TypeConversionExpr)) {
             ((AbstractExpression) rExpr).setMultiReturnAvailable(true);
             rExpr.accept(this);
+            if (assignStmt.isVarDeclaration()) {
+                if (rExpr instanceof TypeCastExpression) {
```

Comment:
Why do you need this if condition? The 'getTypes' method comes from the "ExecutableMultiReturnExpr" interface and both 'FunctionInvocationExpr' and 'ActionInvocationExpr' classes implements that interface.